File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {  
    int dp[101][10001];
    
private:  
    int findMinSquares(int index, int curr_sum, int n, vector<int>& perfectSquares) {
        if(index >= perfectSquares.size()) return 1e9;
        if(curr_sum == n) return 0;
        
        if(dp[index][curr_sum] != -1) return dp[index][curr_sum];
        
        int not_take = findMinSquares(index + 1, curr_sum, n, perfectSquares);
        int take = 1e9;
        if(curr_sum + perfectSquares[index] <= n) {
            take = min(not_take, 1 + findMinSquares(index, curr_sum + perfectSquares[index], n, perfectSquares));
        }
        
        return dp[index][curr_sum] = min(take, not_take);
    }
    
public:
    int numSquares(int n) {
        vector<int> perfectSquares;
        for(int i = 1; i <= 100; i++) {
            perfectSquares.push_back(i * i);
        }
        
        memset(dp, -1, sizeof(dp));
        return findMinSquares(0, 0, n, perfectSquares);
    }
};

File: 0279-perfect-squares/0279-perfect-squares.java
class Solution {
    int dp[][] = new int[101][10001];
    
    private int findMinSquares(int index, int n, int perfectSquares[]) {
        if(index >= perfectSquares.length) return (int)1e9;
        if(n == 0) return 0;
        
        if(dp[index][n] != -1) return dp[index][n];
        
        int not_take = findMinSquares(index + 1, n, perfectSquares);
        int take = (int)1e9;
        if(n - perfectSquares[index] >= 0) {
            take = Math.min(not_take, 1 + findMinSquares(index, n - perfectSquares[index], perfectSquares));
        }
        
        return dp[index][n] = Math.min(take, not_take);
    }
    
    public int numSquares(int n) {
        int perfectSquares[] = new int[100];
        for(int i = 1; i <= 100; i++) {
            perfectSquares[i - 1] = i * i;
        }
        
        for(int i = 0; i < dp.length; i++) {
            Arrays.fill(dp[i], -1);
        }
        
        return findMinSquares(0, n, perfectSquares);
    }
}
